runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MonitoringInfoMetricName.java

```diff
@@ -85,12 +87,33 @@ public String getUrn() {
   }
 
   /** @return The labels associated with this MonitoringInfo. */
-  public HashMap<String, String> getLabels() {
+  public Map<String, String> getLabels() {
     return this.labels;
   }
 
-  public static MonitoringInfoMetricName named(String urn, HashMap<String, String> labels) {
-    return new MonitoringInfoMetricName(urn, labels);
+  /**
+   * Polymorphic constructor of {@link MetricName}s.
+   *
+   * <p>If `urn` is a {@link SimpleMonitoringInfoBuilder#USER_COUNTER_URN_PREFIX metric}, return a
+   * {@link MetricName} auto-value (which will {@link Object#equals equal} and {@link
+   * Object#hashCode hash} consistently with other {@link MetricName}s.
+   *
+   * <p>Otherwise, return a concrete {@link MonitoringInfoMetricName} representing a "system"
+   * metric.
+   */
+  public static MetricName named(String urn, Map<String, String> labels) {
+    checkArgument(!Strings.isNullOrEmpty(urn), "MonitoringInfoMetricName urn must be non-empty");
```

Comment:
I can't think of a reason that this will ever be called with a user beam URN that needs to be parsed. Are you using this somewhere? Otherwise I don't see a reason to add this.

Reply:
In the MonitoringInfos-everywhere world, it's natural to have MIs that are a mix of user- and system-metrics, and want to turn them into SDK reprs (incl. MetricNames).

In this PR, I rely on [the ability to correctly make an `AutoValue_MetricName` for user-metrics and a `MonitoringInfoMetricName` for system metrics in two places]:
- Flink's bundle-progress listener converts mixed MIs [here]().
- The MetricResultsProtos conversion code-path [uses them](https://github.com/apache/beam/pull/7934/files#diff-e722f484e12b515353e20532d3550465R178) ([`keyFromMonitoringInfo`](https://github.com/apache/beam/pull/7934/files#diff-f0ebadd9b3dbb03171ac530fab654d5dR107) calls [`MIMN.create` below](https://github.com/apache/beam/pull/7934/files#diff-ba1b936a7d4c8d43789e81a816b9bb90R115), which calls this); this is how [JobServicePipelineResult to convert MetricResults proto to an SDK MetricResults](https://github.com/apache/beam/pull/7934/files#diff-284acf2526fcf07c05ee44aefbd9d0c3R134).